### ISSUE TITLE
Consider optimizing the square function

### ISSUE THREAD

**Issue by weijiguo** [mit-plv/fiat-crypto#1104]
Word-by-word square could run faster than multiplication, in that we can save some word-size multiplications due to the symmetry: multiplying field elements with 4 limbs requires 16 multiplications: ai * bi for i running from 0 to 3 in (a0, a1, a2, a3) * (b0, b1, b2, b3). But when we do square, a0 * b1 equals a1 * b0. We can save 6 mults, and associated calculations etc. 

The performance improvement is not trivial. Consider the [P-256 for 64 bit target in Golang](https://github.com/mit-plv/fiat-crypto/blob/master/fiat-go/64/p256/p256.go) , the currently generated code requires 28 mults. If we cut 6 mults and associated calculations we could gain about a quarter more performance for square. Then this will help the field inversion a lot if we use [addchain](https://github.com/mmcloughlin/addchain) , which gives mostly squares in its plan. For 32 bit target, the save is 28 mults out of 96.

**Comment by JasonGross** [comment]
We optimize square for unsaturated solinas at https://github.com/mit-plv/fiat-crypto/blob/5f4145adb764af03d787fbe3906218ce3f8adc51/src/Arithmetic/Core.v#L59-L70 (as compared with mul at https://github.com/mit-plv/fiat-crypto/blob/5f4145adb764af03d787fbe3906218ce3f8adc51/src/Arithmetic/Core.v#L50-L54)
However, the definition of word-by-word mul is much more complicated:
https://github.com/mit-plv/fiat-crypto/blob/5f4145adb764af03d787fbe3906218ce3f8adc51/src/Arithmetic/WordByWordMontgomery.v#L59-L111
https://github.com/mit-plv/fiat-crypto/blob/5f4145adb764af03d787fbe3906218ce3f8adc51/src/Arithmetic/WordByWordMontgomery.v#L1043-L1044

In particular, it's not clear to me how to take advantage of symmetry in
https://github.com/mit-plv/fiat-crypto/blob/5f4145adb764af03d787fbe3906218ce3f8adc51/src/Arithmetic/WordByWordMontgomery.v#L72-L81
If you have a suggestion of how to adjust the general algorithm (or even better, a PR), I'd be happy to (help) integrate that optimization.

**Comment by weijiguo** [comment]
IMHO the symmetry information has been lost when defining (squaremod a) as (mulmod a a): redc has no way to tell if a equals b in its parameter list.

https://github.com/mit-plv/fiat-crypto/blob/5f4145adb764af03d787fbe3906218ce3f8adc51/src/Arithmetic/WordByWordMontgomery.v#L1043-L1044

Is it possible that we have another reduction algorithm for squaring? Sth like:

`
  Definition squaremod (a : list Z) : list Z := @redc_s bitwidth n m_enc n a m'.  
`

Then **redc_s** can try to use saved earlier computations whenever possible.

I have no idea how this could be achieved in Coq, which is a total stranger to me. BTW I just applied the symmetry trick to generated code to save some multiplications, the performance gain is only marginal (~ 2%), opposing the earlier guess.

**Comment by JasonGross** [comment]
> IMHO the symmetry information has been lost when defining `squaremod a` as `mulmod a a`: `redc` has no way to tell if `a` equals `b` in its parameter list.

Indeed, the thing to do is define a variant of `redc` which takes one argument rather than two, and prove something like `redc_square a = redc a a`

> Edited: With "regular" programming languages, generally **redc** can try to use saved earlier computations whenever possible: first populating a 2-dimension multiplication table `t` of `ai*bj` such that
> 
> `t[i][j] = ai * bj`
> 
> `redc` then looks up table entries when a product is needed. `mulmod` and `squaremod` can have different ways to populate table `t`.

Indeed, this is possible, though quite cumbersome.  The way to do this in Coq is roughly the same as the way to do it in Haskell: define a variant of the state monad whose only nontrivial function is "get the product ai * bj as a function of i and j".  This can then be initialized either with a and b or with two copies of a, and the method will update the threaded state as necessary.

That is, we define:
```coq
Definition State T := list ((nat * nat) * T).
Definition M T := State T -> T * State T.
Definition bind {A B} : M A -> (A -> M B) -> M B := ...
Definition ret {A} (a : A) : M A := fun st => (a, st).
Definition run {A} (f : M A) : A := fst (f nil).
```
And then we parameterize our functions over `memo_mul : nat -> nat -> M Z` which will, depending on if it's mul or square, memoize potentially modulo swapping the indices.  

This is a pretty clunky solution, though, and quite invasive.

The thing I was asking about was a way of knowing statically (before you even know the size of the prime and the number of limbs) how to write down the version of the algorithm that doesn't need table lookups to deduplicate multiplication, and is instead clever about multiplying by 2 in the right places and managing list lengths.  It was clear how to do this when multiplication was just "add all the ways of multiplying an element in the first list with an element in the second"; instead we can take each element `t` from the list, add `t^2` to twice `t` times the sum of the elements in the list after `t`, and sum this result over all the elements in the list.  I'm looking for a similar description of how to modify `redc`.

**Comment by JasonGross** [comment]
Having stared at the code a bit, I think that
https://github.com/mit-plv/fiat-crypto/blob/5f4145adb764af03d787fbe3906218ce3f8adc51/src/Arithmetic/WordByWordMontgomery.v#L76
iterates through the product of each element of A times all of B.  I wonder if there's a way of reordering the partial products when we're squaring to avoid duplication.  For example, if instead of multiplying a by all of B, we instead started at the index of a, zeroing out elements that came before, and doubling everything after the index of a.

@andres-erbsen @jadephilipoom thoughts on this?

**Comment by JasonGross** [comment]
I guess there's an issue that if we're squaring `2^64-1` and then multiplying by 2, there's no guarantee that this will fit in 128 bits, which means that we'll still end up with the same number of partial products total (and I think this means that we can't actually reorder the partial products like this without risking overflow?)

**Comment by weijiguo** [comment]
>  The thing I was asking about was a way of knowing statically (before you even know the size of the prime and the number of limbs) how to write down the version of the algorithm that doesn't need table lookups to deduplicate multiplication, and is instead clever about multiplying by 2 in the right places and managing list lengths. It was clear how to do this when multiplication was just "add all the ways of multiplying an element in the first list with an element in the second"

I am not sure if multiplying by 2 works best here. We are not obtaining the full product of a * b in Word by Word Montgomery: a<sub>0</sub>*b<sub>1</sub> does not add to a<sub>1</sub>*b<sub>0</sub>, just for example. 

Meanwhile, from the perspective of any target language of this project, it would be beneficial to concentrate the calculation of all such a<sub>i</sub>*b<sub>j</sub> in one place, perhaps its own function, to create an optimization opportunity for vectorization.
